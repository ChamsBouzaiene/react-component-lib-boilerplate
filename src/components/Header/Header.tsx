import React from 'react';
import PropTypes, { string } from 'prop-types';
import Button from '../Button/Button';
import { Icon, Title, Container } from '../atoms';

export interface HeaderProps {
    user?: Record<string, unknown>;
    onLogin: () => void;
    onLogout: () => void;
    onCreateAccount: () => void;
}

const Header: React.FC<HeaderProps> = ({ user, onLogin, onLogout, onCreateAccount }) => (
    <header>
        <Container>
            <div>
                <Icon />
                <Title fontWeight={800} fontSize={20} lineHeight={1} margin="6px 0 6px 10px" verticalAlign="top">
                    Acme
                </Title>
            </div>
            <div>
                {user ? (
                    <Button size="small" onClick={onLogout} label="Log out" />
                ) : (
                    <>
                        <Button size="small" onClick={onLogin} label="Log in" />
                        <Button
                            variant="primary"
                            size="small"
                            marginLeft={10}
                            onClick={onCreateAccount}
                            label="Sign up"
                        />
                    </>
                )}
            </div>
        </Container>
    </header>
);

Header.propTypes = {
    user: PropTypes.objectOf(string),
    onLogin: PropTypes.func.isRequired,
    onLogout: PropTypes.func.isRequired,
    onCreateAccount: PropTypes.func.isRequired,
};

export default Header;
