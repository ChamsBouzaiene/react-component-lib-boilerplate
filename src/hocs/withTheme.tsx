import React from 'react';
import { ThemeProvider } from 'styled-components';
import darkTheme from '../theme';

export function withTheme<P>(WrappedComponent: React.ComponentType<P>): React.FC<P> {
    const ComponentWithTheme = (props: P) => {
        return (
            <ThemeProvider theme={darkTheme}>
                <WrappedComponent {...props} />
            </ThemeProvider>
        );
    };
    return ComponentWithTheme;
}
export default withTheme;
