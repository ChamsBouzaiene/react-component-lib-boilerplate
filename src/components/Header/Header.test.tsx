import React from 'react';
import { render, screen } from '@testing-library/react';
import Header from './Header';
import { withTheme } from '../../hocs/withTheme';
import userEvent from '@testing-library/user-event';

const ThemedHeader = withTheme(Header);

describe('Button', () => {
    test('renders a default header while user is underfined', () => {
        const mockOnLogin = jest.fn();
        const mockOnLogout = jest.fn();
        const mockOnCreateAccount = jest.fn();

        const { baseElement } = render(
            <ThemedHeader onLogin={mockOnLogin} onLogout={mockOnLogout} onCreateAccount={mockOnCreateAccount} />,
        );
        const loginButton = screen.queryByText('Log in');
        const logoutButton = screen.queryByText('Log out');
        const signupButton = screen.queryByText('Sign up');

        expect(screen.getByText('Acme')).toBeInTheDocument();
        expect(loginButton).toBeInTheDocument();
        expect(logoutButton).not.toBeInTheDocument();
        expect(signupButton).toBeInTheDocument();
        if (loginButton) {
            userEvent.click(loginButton);
            expect(mockOnLogin).toHaveBeenCalledTimes(1);
        }
        if (signupButton) {
            userEvent.click(signupButton);
            expect(mockOnCreateAccount).toHaveBeenCalledTimes(1);
        }

        expect(baseElement).toMatchSnapshot();
    });
    test('renders a default header while user is defined ', () => {
        const mockOnLogin = jest.fn();
        const mockOnLogout = jest.fn();
        const mockOnCreateAccount = jest.fn();

        const { baseElement } = render(
            <ThemedHeader
                user={{ name: 'chams' }}
                onLogin={mockOnLogin}
                onLogout={mockOnLogout}
                onCreateAccount={mockOnCreateAccount}
            />,
        );
        const LoginButton = screen.queryByText('Log in');
        const LogoutButton = screen.queryByText('Log out');
        const signupButton = screen.queryByText('Sign up');

        expect(screen.getByText('Acme')).toBeInTheDocument();
        expect(LoginButton).not.toBeInTheDocument();
        expect(LogoutButton).toBeInTheDocument();
        expect(signupButton).not.toBeInTheDocument();

        if (LogoutButton) {
            userEvent.click(LogoutButton);
            expect(mockOnLogout).toHaveBeenCalledTimes(1);
        }

        expect(baseElement).toMatchSnapshot();
    });
});
