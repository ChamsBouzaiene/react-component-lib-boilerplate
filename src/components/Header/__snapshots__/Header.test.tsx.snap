// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders a default header while user is defined  1`] = `
<body>
  <div>
    <header>
      <div
        class="sc-dlnjwi czjjER"
      >
        <div>
          <svg
            class="sc-gtsrHT fNKQCe"
            height="32"
            viewBox="0 0 32 32"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="M10 0h12a10 10 0 0110 10v12a10 10 0 01-10 10H10A10 10 0 010 22V10A10 10 0 0110 0z"
                fill="#FFF"
              />
              <path
                d="M5.3 10.6l10.4 6v11.1l-10.4-6v-11zm11.4-6.2l9.7 5.5-9.7 5.6V4.4z"
                fill="#555AB9"
              />
              <path
                d="M27.2 10.6v11.2l-10.5 6V16.5l10.5-6zM15.7 4.4v11L6 10l9.7-5.5z"
                fill="#91BAF8"
              />
            </g>
          </svg>
          <h1
            class="sc-hKFxyN fFmBQZ"
            font-size="20"
            font-weight="800"
          >
            Acme
          </h1>
        </div>
        <div>
          <button
            class="sc-bdnxRM dTEMSW"
          >
            Log out
          </button>
        </div>
      </div>
    </header>
  </div>
</body>
`;

exports[`Button renders a default header while user is underfined 1`] = `
<body>
  <div>
    <header>
      <div
        class="sc-dlnjwi czjjER"
      >
        <div>
          <svg
            class="sc-gtsrHT fNKQCe"
            height="32"
            viewBox="0 0 32 32"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="M10 0h12a10 10 0 0110 10v12a10 10 0 01-10 10H10A10 10 0 010 22V10A10 10 0 0110 0z"
                fill="#FFF"
              />
              <path
                d="M5.3 10.6l10.4 6v11.1l-10.4-6v-11zm11.4-6.2l9.7 5.5-9.7 5.6V4.4z"
                fill="#555AB9"
              />
              <path
                d="M27.2 10.6v11.2l-10.5 6V16.5l10.5-6zM15.7 4.4v11L6 10l9.7-5.5z"
                fill="#91BAF8"
              />
            </g>
          </svg>
          <h1
            class="sc-hKFxyN fFmBQZ"
            font-size="20"
            font-weight="800"
          >
            Acme
          </h1>
        </div>
        <div>
          <button
            class="sc-bdnxRM dTEMSW"
          >
            Log in
          </button>
          <button
            class="sc-bdnxRM fVYaCq"
          >
            Sign up
          </button>
        </div>
      </div>
    </header>
  </div>
</body>
`;
